### PR TITLE
[IMP] pos: Change waiter screen

### PR DESCRIPTION
**Pull request by adgu-odoo**
The PoS needed improvments for the mobile use.
This commit is changing the display of some screen and popups like the NumberPopup. It also changes the logic of the order count badge on the table. It now displays the number of orderline (and their quantity) to send to the printers or preparation display if there is at least one.

Enterprise PR: odoo/enterprise#41013
task-id: 2893634

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/pos_restaurant/static/src/app/floor_screen/table.js [hunk 1]
```diff
@@ -70,11 +90,13 @@ export class Table extends Component {
...
-            : this.pos.globalState
-                  .getTableOrders(table.id)
-                  .filter((o) => o.orderlines.length !== 0 || o.paymentlines.length !== 0).length;
+
+        if (table.changes_count > 0) {
+            return table.changes_count;
+        } else if (table.skip_changes > 0) {
+            return table.skip_changes;
+        }
+        return table.order_count;
```
Comment (davidmonnom) [inline addons/pos_restaurant/static/src/app/floor_screen/table.js:99]: With the move of ` await this.pos.globalState._syncTableOrdersToServer();` this function should maybe be modified to pass tests. Will see..

File: addons/pos_restaurant/static/src/js/Misc/MobileOrderWidget.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        const order = this.pos.globalState.get_order();
+        const orderline = order.get_selected_orderline();
+        if (this.pos.globalState.config.module_pos_restaurant && this.pos.globalState.orderPreparationCategories.size) {
+            const changes = order.getOrderChanges().orderlines;
+            for (const change of Object.values(changes)) {
+                if (change.product_id == orderline.product.id) {
+                    return change.quantity;
+                }
+            }
+            return false;
```
Comment (davidmonnom) [inline addons/pos_restaurant/static/src/js/Misc/MobileOrderWidget.js:20]:
```suggestion
            const changes = Object.values(order.getOrderChanges().orderlines).find(
                (change) => change.product_id == orderline.product.id
            );
            return changes ? changes.quantity : false;
```

File: addons/pos_restaurant/static/src/js/Misc/MobileOrderWidget.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            this.pos.globalState.orderPreparationCategories.size
+        );
+    },
+    submitOrder() {
+        this.pos.sendOrderInPreparation(this.pos.globalState.get_order());
+    },
+    primaryPayButton() {
+        return !this.order.is_empty() && ((!this.swapButton && this._super(...arguments)) || (this.swapButton && this.pos.globalState.get_order().getOrderChanges().count > 0));
+    },
+});
```
Comment (davidmonnom) [inline addons/pos_restaurant/static/src/js/Misc/MobileOrderWidget.js:39]: Thank you for allowing me to write **EOF** for the first time.

File: addons/pos_restaurant/static/src/js/models.js [hunk 4]
```diff
@@ -157,7 +157,6 @@ patch(PosGlobalState.prototype, "pos_restaurant.PosGlobalState", {
         return this.get_order_list().filter((order) => order.tableId === tableId);
     },
     unsetTable() {
-        this._syncTableOrdersToServer();
```
Comment (davidmonnom) [inline addons/pos_restaurant/static/src/js/models.js:160]: Same comment as above, this line should be reinstated.

File: addons/pos_restaurant/tests/test_frontend.py [hunk 5]
```diff
@@ -47,6 +54,7 @@ def setUp(self):
             'iface_orderline_notes': True,
             'iface_start_categ_id': drinks_category.id,
             'start_category': True,
+            'is_order_printer': True,
```
Comment (davidmonnom) [inline addons/pos_restaurant/tests/test_frontend.py:57]: Is this usefull ? Its never used in frontend

**Review by davidmonnom — COMMENTED** [review]
Here are some points that I couldn't put in the review.

I see that the size of the buttons "validate", "pay", "continue", "new order" has changed in different views. This may be related to one of the points in your task: 

image

---

Could you ask Sam how to set up the linter? So that you don't get useless diffs when others lint the whole file?

Thx

File: addons/point_of_sale/static/src/js/Screens/PaymentScreen/PaymentScreen.js [hunk 6]
```diff
@@ -59,15 +59,6 @@ export class PaymentScreen extends Component {
...
         };
-        // Check if pos has a cash payment method
-        const hasCashPaymentMethod = this.payment_methods_from_config.some(
-            (method) => method.type === "cash"
-        );
-
-        if (!hasCashPaymentMethod) {
-            config["maxValue"] = this.currentOrder.get_due();
-            config["maxValueReached"] = this.showMaxValueError.bind(this);
-        }
```
[the remark this answers is not in the document]
Comment (adgu-odoo) [reply addons/point_of_sale/static/src/js/Screens/PaymentScreen/PaymentScreen.js:70]: It is ok this maxValue is related to the maximum value that you can put in a payment line and the logic is just moved elsewhere in the code.

File: addons/point_of_sale/static/src/js/Screens/TicketScreen/TicketScreen.js [hunk 7]
```diff
@@ -279,7 +279,7 @@ export class TicketScreen extends IndependentToOrderScreen {
             globalState.set_order(destinationOrder);
         }
 
-        this.onCloseScreen();
+        await this.onCloseScreen();
```
Comment (davidmonnom) [inline addons/point_of_sale/static/src/js/Screens/TicketScreen/TicketScreen.js:282]: Same :)

File: addons/point_of_sale/static/src/js/models.js [hunk 8]
```diff
@@ -3093,14 +3098,20 @@ export class Order extends PosModel {
...
+                if (quantityDiff && orderline.skipChange === skipped) {
+                    const diff = {
                         name: orderline.get_full_product_name(),
                         product_id: product.id,
                         quantity: quantityDiff,
                         note: note,
                     };
-                    orderline.setHasChange(true);
+
+                    changes[productKey] = diff;
```
Comment (davidmonnom) [inline addons/point_of_sale/static/src/js/models.js:3109]: Why const declaration ?

File: addons/point_of_sale/static/src/xml/Screens/ProductScreen/ProductScreen.xml [hunk 9]
```diff
@@ -22,25 +19,31 @@
...
-                            <h1>Pay</h1>
-                            <span t-esc="total"/>
-                        </button>
-                        <button class="btn-switchpane secondary" t-on-click="switchPane">
-                            <h1>Review</h1>
-                            <span><t t-esc="items"/> items</span>
-                        </button>
-                    </div>
+                    <t t-if="ui.isSmall">
+                        <div class="product-reminder" t-if="showProductReminder" t-key="animationKey" t-ref="reminder">
```
Comment (davidmonnom) [inline addons/point_of_sale/static/src/xml/Screens/ProductScreen/ProductScreen.xml:32]: Same, in other PR remove the ref

File: addons/point_of_sale/static/src/xml/Screens/ReceiptScreen/ReceiptScreen.xml [hunk 10]
```diff
@@ -65,6 +65,24 @@
...
+                    <div class="btn-switchpane validation-button" t-att-class="{ highlight: !locked }"
+                        t-if="!splittedOrder"
+                        t-on-click="orderDone" name="done">
+                        New Order
+                    </div>
+                    <div t-if="isResumeVisible() and !splittedOrder" class="btn-switchpane validation-button" t-att-class="{ highlight: !locked }"
+                        t-on-click="resumeOrder" name="resume">
+                        Resume Order
+                    </div>
+                    <div t-if="splittedOrder" class="button next validation" t-att-class="{ highlight: !locked }"
```
Comment (davidmonnom) [inline addons/point_of_sale/static/src/xml/Screens/ReceiptScreen/ReceiptScreen.xml:78]: Change "next validation" to "btn-switchpane validation-button" (in other pR)

**Review by davidmonnom — COMMENTED** [review]
Some change to make in another PR like that we can merge this one !

**Comment by davidmonnom** [comment]
@robodoo r+

**Pull request closed without merging**